src/gc/env/gcenv.object.h

```diff
@@ -64,19 +68,31 @@ class MethodTable
         return m_componentSize;
     }
 
+    bool Collectible()
+    {
+        return (m_flags & MTFlag_Collectible) != 0;
+    }
+
     bool ContainsPointers()
     {
         return (m_flags & MTFlag_ContainsPointers) != 0;
     }
 
     bool ContainsPointersOrCollectible()
     {
-        return ContainsPointers();
+        return ContainsPointers() || Collectible();
     }
 
     bool HasComponentSize()
     {
-        return m_componentSize != 0;
+        // Note that we can't just check m_componentSize != 0 here. The VM 
+        // may still construct a method table that does not have a component 
+        // size, according to this method, but still has a number in the low 
+        // 16 bits of the method table flags parameter.
+        //
+        // The solution here is to do what the VM does and check the
+        // HasComponentSize flag so that we're on the same page.
+        return (m_flags & MTFlag_HasComponentSize) != 0;
```

Comment:
This was hard to track down - it should be accurate now with what the VM does with the m_flags field. CoreRT doesn't use this code path (it forwards methods on the MethodTable facade to `EEType`).